Add Sanyo A/C (72 bit) protocol with detailed support.

crankyoldgit opened the pull request:
* Add basic send and decode unit tests for new protocol.
* Add various protocol settings:
  - Power
  - Desired Temp
  - Operating Mode
  - Fan Speed
  - Vertical Swing
  - Sleep / Night setback
  - Off Timer
  - Temp Sensor Location & Value
  - Checksum
* Add to Common A/C API.
* Unit tests for Sanyo A/C

Fixes #1211

crankyoldgit committed:
Sanyo A/C: Basic send/decode support
* `sendSanyoAc()` & `decodeSanyoAc()` support
* Update unit tests.
* Add basic send and decode unit tests for new protocol.

For #1211

crankyoldgit committed:
Tune the Sanyo A/C timings
* Per https://github.com/crankyoldgit/IRremoteESP8266/issues/1211#issuecomment-656536444

crankyoldgit committed:
Initial Sanyo A/C Class
* Add checksum support.

crankyoldgit committed:
SanyoAc: Minimal detailed support
* Add Power, Mode, Temp, & Fan speed settings.
* Add to Common A/C API.
* Unit tests for Power, Mode, Temp, & Fan speed.
* Update existing Unit tests.

* Misc IRac style cleanup.

For #1211

crankyoldgit committed:
SanyoAc: Add Sleep (Night Setback) setting.
For #1211

crankyoldgit committed:
SanyoAc: Add Vertical Swing Support
* Add vertical Swing setting support.
* This should be the last additional setting for the Common AC API support. The remainder are `IRSanyoAc` class only.

For #1211

crankyoldgit committed:
SanyoAc: Add support for the sensor location & temp.
* Refactor set & get functions for temp.
* Add Sensor (location) and Sensor Temp.
* Default to "Wall"/"AC Unit" for sensor location for Common A/C API. Also use desired temp as sensor temp.

For #1211

crankyoldgit committed:
SanyoAc: Add support for Off Timer.
Ref #1211

NiKiZe commented on src/ir_Sanyo.cpp:
```diff
@@ -221,3 +240,440 @@ bool IRrecv::decodeSanyo(decode_results *results, uint16_t nbits, bool strict) {
+bool IRrecv::decodeSanyoAc(decode_results *results, uint16_t offset,
+                           const uint16_t nbits, const bool strict) {
+  if (strict && nbits != kSanyoAcBits)
+    return false;
```
No checking of length?

crankyoldgit replied:
Nope. If a decode routine solely uses a single `matchGeneric()` then there is no need to do a length check as it does one. See:
https://github.com/crankyoldgit/IRremoteESP8266/blob/5a007a81ad33684346a5a19c4dd85508d9bc9bf1/src/IRrecv.cpp#L1192-L1201

NiKiZe replied:
I see, thanks

crankyoldgit committed:
SanyoAc: Add support for controling the Beep setting.

Fixes #1211

crankyoldgit committed:
SanyoAc: Update to Stable.
Reported as working in https://github.com/crankyoldgit/IRremoteESP8266/issues/1211#issuecomment-657178313

crankyoldgit commented:
Confirmed working in https://github.com/crankyoldgit/IRremoteESP8266/issues/1211#issuecomment-657178313

NiKiZe commented on test/ir_Sanyo_test.cpp:
```diff
@@ -257,3 +261,267 @@ TEST(TestDecodeSanyoLC7461, FailToDecodeNonSanyoLC7461Example) {
+  IRSanyoAc ac(kGpioUnused);
+  ac.begin();
+
+  ac.setOffTimer(0);
+  EXPECT_EQ(0, ac.getOffTimer());
+  ac.setOffTimer(59);
+  EXPECT_EQ(0, ac.getOffTimer());
+  ac.setOffTimer(60);
+  EXPECT_EQ(60, ac.getOffTimer());
+  ac.setOffTimer(61);
+  EXPECT_EQ(60, ac.getOffTimer());
+  ac.setOffTimer(15 * 60 + 59);
+  EXPECT_EQ(15 * 60, ac.getOffTimer());
+  ac.setOffTimer(16 * 60);
+  EXPECT_EQ(15 * 60, ac.getOffTimer());
+
+  const uint8_t offTimer2Hr[kSanyoAcStateLength] = {
+      0x6A, 0x6D, 0x4F, 0x02, 0x14, 0x85, 0x00, 0x00, 0x4A};
+  ac.setRaw(offTimer2Hr);
+  EXPECT_EQ(2 * 60, ac.getOffTimer());
```
No `resultAcToString` tests of off timer data

crankyoldgit replied:
Good catch. Yes, I was lazy. Fixed/done.

crankyoldgit committed:
Code Review Feedback
* Add `toString()` unit test coverage for the Off Timer.
For #1211

crankyoldgit commented:
@NiKiZe Thanks (as usual!) for the quick and thorough code review!
